### ISSUE TITLE
Karpenter can not deprovision node

### ISSUE THREAD

**Issue by hitsub2** [aws/karpenter#3794]
Karpenter did not deprovision GPU node when non-daemonset is running.

#### Pod info
```
$ kubectl get pod --field-selector spec.nodeName=[ip-172-24-44-111.us](http://ip-172-24-44-111.us/)-west-2.compute.internal -A                                              

NAMESPACE     NAME                 READY   STATUS    RESTARTS   AGE

default       fsx-csi-node-9994v   3/3     Running   0          6h17m

kube-system   aws-node-g9c4n       1/1     Running   0          6h16m

kube-system   kube-proxy-xpzfp     1/1     Running   0          6h17m

 
$ kubectl get daemonset -A                                                                                                              1 ↵  

NAMESPACE     NAME           DESIRED   CURRENT   READY   UP-TO-DATE   AVAILABLE   NODE SELECTOR            AGE

default       fsx-csi-node   16        16        16      16           16          [kubernetes.io/os=linux](http://kubernetes.io/os=linux)   6d21h

kube-system   aws-node       16        16        16      16           16          <none>                   6d22h

kube-system   kube-proxy     16        16        16      16           16          <none>                   6d22h

```

#### karpenter provisioner

```python

apiVersion: karpenter.sh/v1alpha5
kind: Provisioner
metadata:
  generation: 1
  name: gpu-karpenter
spec:
  labels:
    instance-group: gpu-group
  limits:
    resources:
      cpu: 1k
  providerRef:
    name: gpu-karpenter
  requirements:
    - key: [node.kubernetes.io/instance-type](http://node.kubernetes.io/instance-type)
      operator: In
      values:
        - g3.4xlarge
        - p3.2xlarge
    - key: [kubernetes.io/os](http://kubernetes.io/os)
      operator: In
      values:
        - linux
    - key: [kubernetes.io/arch](http://kubernetes.io/arch)
      operator: In
      values:
        - amd64
    - key: karpenter.sh/capacity-type
      operator: In
      values:
        - on-demand
  ttlSecondsAfterEmpty: 30
```

#### Node info

```
apiVersion: v1
kind: Node
metadata:
  name: ip-172-24-37-21.us-west-2.compute.internal
  uid: b92594a7-8f79-4ff2-919a-3e6fa53bad1f
  resourceVersion: ‘3398073’
  creationTimestamp: ‘2023-04-21T04:34:59Z’
  labels:
    [beta.kubernetes.io/arch](http://beta.kubernetes.io/arch): amd64
    [beta.kubernetes.io/instance-type](http://beta.kubernetes.io/instance-type): g3.4xlarge
    [beta.kubernetes.io/os](http://beta.kubernetes.io/os): linux
    [failure-domain.beta.kubernetes.io/region](http://failure-domain.beta.kubernetes.io/region): us-west-2
    [failure-domain.beta.kubernetes.io/zone](http://failure-domain.beta.kubernetes.io/zone): us-west-2c
    instance-group: gpu-group
    [k8s.io/cloud-provider-aws](http://k8s.io/cloud-provider-aws): 6ed6dc4edaedb5f6ef7a3f14ed1cebca
    karpenter.k8s.aws/instance-ami-id: ami-0c9ca2e944fa15a23
    karpenter.k8s.aws/instance-category: g
    karpenter.k8s.aws/instance-cpu: ‘16’
    karpenter.k8s.aws/instance-encryption-in-transit-supported: ‘false’
    karpenter.k8s.aws/instance-family: g3
    karpenter.k8s.aws/instance-generation: ‘3’
    karpenter.k8s.aws/instance-gpu-count: ‘1’
    karpenter.k8s.aws/instance-gpu-manufacturer: nvidia
    karpenter.k8s.aws/instance-gpu-memory: ‘8192’
    karpenter.k8s.aws/instance-gpu-name: m60
    karpenter.k8s.aws/instance-hypervisor: xen
    karpenter.k8s.aws/instance-memory: ‘124928’
    karpenter.k8s.aws/instance-pods: ‘234’
    karpenter.k8s.aws/instance-size: 4xlarge
    karpenter.sh/capacity-type: on-demand
    karpenter.sh/provisioner-name: gpu-karpenter
    [kubernetes.io/arch](http://kubernetes.io/arch): amd64
    [kubernetes.io/hostname](http://kubernetes.io/hostname): ip-172-24-37-21.us-west-2.compute.internal
    [kubernetes.io/os](http://kubernetes.io/os): linux
    [node.kubernetes.io/instance-type](http://node.kubernetes.io/instance-type): g3.4xlarge
    [topology.kubernetes.io/region](http://topology.kubernetes.io/region): us-west-2
    [topology.kubernetes.io/zone](http://topology.kubernetes.io/zone): us-west-2c
  annotations:
    [alpha.kubernetes.io/provided-node-ip](http://alpha.kubernetes.io/provided-node-ip): 172.24.37.21
    [csi.volume.kubernetes.io/nodeid](http://csi.volume.kubernetes.io/nodeid): ‘{“[fsx.csi.aws.com](http://fsx.csi.aws.com/)”:“i-092cc6c0b3032347e”}’
    [node.alpha.kubernetes.io/ttl](http://node.alpha.kubernetes.io/ttl): ‘0’
    [volumes.kubernetes.io/controller-managed-attach-detach](http://volumes.kubernetes.io/controller-managed-attach-detach): ‘true’
  ownerReferences:
    - apiVersion: karpenter.sh/v1alpha5
      kind: Provisioner
      name: gpu-karpenter
      uid: 88e49adc-981e-4afa-898b-a2ecf9fba6c2
      blockOwnerDeletion: true
  finalizers:
    - karpenter.sh/termination
status:
  capacity:
    attachable-volumes-aws-ebs: ‘39’
    cpu: ‘16’
    ephemeral-storage: 20959212Ki
    hugepages-1Gi: ‘0’
    hugepages-2Mi: ‘0’
    memory: 125690960Ki
    pods: ‘234’
  allocatable:
    attachable-volumes-aws-ebs: ‘39’
    cpu: 15890m
    ephemeral-storage: ‘18242267924’
    hugepages-1Gi: ‘0’
    hugepages-2Mi: ‘0’
    memory: 122691664Ki
    pods: ‘234’
  conditions:
    - type: MemoryPressure
      status: ‘False’
      lastHeartbeatTime: ‘2023-04-23T03:32:45Z’
      lastTransitionTime: ‘2023-04-21T04:35:41Z’
      reason: KubeletHasSufficientMemory
      message: kubelet has sufficient memory available
    - type: DiskPressure
      status: ‘False’
      lastHeartbeatTime: ‘2023-04-23T03:32:45Z’
      lastTransitionTime: ‘2023-04-21T04:35:41Z’
      reason: KubeletHasNoDiskPressure
      message: kubelet has no disk pressure
    - type: PIDPressure
      status: ‘False’
      lastHeartbeatTime: ‘2023-04-23T03:32:45Z’
      lastTransitionTime: ‘2023-04-21T04:35:41Z’
      reason: KubeletHasSufficientPID
      message: kubelet has sufficient PID available
    - type: Ready
      status: ‘True’
      lastHeartbeatTime: ‘2023-04-23T03:32:45Z’
      lastTransitionTime: ‘2023-04-21T04:36:13Z’
      reason: KubeletReady
      message: kubelet is posting ready status
  addresses:
    - type: InternalIP
      address: 172.24.37.21
    - type: Hostname
      address: ip-172-24-37-21.us-west-2.compute.internal
    - type: InternalDNS
      address: ip-172-24-37-21.us-west-2.compute.internal
  daemonEndpoints:
    kubeletEndpoint:
      Port: 10250
  nodeInfo:
    machineID: 728cad23530447f68a1d1000418d0e2d
    systemUUID: ec248bef-7141-eaa5-f758-2df20abebcd7
    bootID: 90d7db47-ce6b-4e87-802c-d8d202328626
    kernelVersion: 5.4.238-148.347.amzn2.x86_64
    osImage: Amazon Linux 2
    containerRuntimeVersion: containerd://1.6.19
    kubeletVersion: v1.24.11-eks-a59e1f0
    kubeProxyVersion: v1.24.11-eks-a59e1f0
    operatingSystem: linux
    architecture: amd64
```

#### karpenter logs

```
023-04-21T06:05:00.044Z	INFO	controller.inflightchecks	Inflight check failed for node, Expected resource "nvidia.com/gpu" didn't register on the node	{"commit": "beb0a64-dirty", "node": "ip-172-24-37-21.us-west-2.compute.internal"}
2023-04-21T06:14:22.629Z	INFO	controller.inflightchecks	Inflight check failed for node, Expected resource "nvidia.com/gpu" didn't register on the node	{"commit": "beb0a64-dirty", "node": "ip-172-24-33-78.us-west-2.compute.internal"}
2023-04-21T06:14:22.638Z	INFO	controller.inflightchecks	Inflight check failed for node, Expected resource "nvidia.com/gpu" didn't register on the node	{"commit": "beb0a64-dirty", "node": "ip-172-24-36-10.us-west-2.compute.internal"}
2023-04-21T06:14:22.646Z	INFO	controller.inflightchecks	Inflight check failed for node, Expected resource "nvidia.com/gpu" didn't register on the node	{"commit": "beb0a64-dirty", "node": "ip-172-24-43-174.us-west-2.compute.internal"}
2023-04-21T06:14:22.651Z	INFO	controller.inflightchecks	Inflight check failed for node, Expected resource "nvidia.com/gpu" didn't register on the node	{"commit": "beb0a64-dirty", "node": "ip-172-24-36-238.us-west-2.compute.internal"}
2023-04-21T06:14:22.663Z	INFO	controller.inflightchecks	Inflight check failed for node, Expected resource "nvidia.com/gpu" didn't register on the node	{"commit": "beb0a64-dirty", "node": "ip-172-24-38-153.us-west-2.compute.internal"}
2023-04-21T06:14:22.664Z	INFO	controller.inflightchecks	Inflight check failed for node, Expected resource "nvidia.com/gpu" didn't register on the node
```

#### version info

karpenter 0.25.0
eks 1.24.0

**Comment by hitsub2** [comment]
According to the [docs](https://karpenter.sh/preview/troubleshooting/#expected-resources-are-registered), Common resources that don’t register and leave nodes in a non-initialized state, but the node is ready.

**Comment by korenyoni** [comment]
Looks like this is related to https://github.com/aws/karpenter/issues/2021

**Comment by jonathan-innis** [comment]
There's some context into the initialization logic [here](https://github.com/aws/karpenter/issues/2833#issuecomment-1505601649).

> common resources didn't register

As @korenyoni mentioned, this behavior is changing in `v0.28.0` to only consider the "requested resources" from the pods that are planning to be scheduled on the launched machine for initialization. This means, like mentioned in the context of the issue linked above, that if you use an instance type that has a known extended resource type but don't install the DS plugin, the node will still be considered initialized, so long as the pods that schedule to that node don't have that extended resource in their pod requests.

**Comment by jonathan-innis** [comment]
I'm going to close this as a duplicate of #2833, noting that that issue will be closed when `v0.28.0` is relased.
